otelx/tracing.go

```diff
@@ -0,0 +1,222 @@
+	//
+	//      tracing.otlp.endpoint        TRACING_OTLP_ENDPOINT        url for otlp http endpoint
+	//      tracing.otlp.insecure        TRACING_OTLP_INSECURE        use an insecure connection to the endpoint
+	//      tracing.otlp.timeout         TRACING_OTLP_TIMEOUT         timeout for sending to the endpoint (defaults to 10s)
+	ExporterOTLPHTTP TraceExporter = "otlphttp"
+
+	// ExporterOTLPGRPC is for a OTLP exporter capable of connecting over secure or
+	// insecure GRPC.
+	//
+	//      tracing.otlp.endpoint        TRACING_OTLP_ENDPOINT        url for otlp http endpoint
+	//      tracing.otlp.insecure        TRACING_OTLP_INSECURE        use an insecure connection to the endpoint
+	//      tracing.otlp.timeout         TRACING_OTLP_TIMEOUT         timeout for sending to the endpoint (defaults to 10s)
+	ExporterOTLPGRPC TraceExporter = "otlpgrpc"
+
+	// ExporterPassthrough is to configure tracing as a passthrough service. This
+	// will setup a tracer and read incoming parent trace info from request and
+	// pass parent trace info to downstream services, but will not export any spans
+	// or other trace data from the application directly.
+	ExporterPassthrough TraceExporter = "passthrough"
+)
+
+var (
+	// ErrUnknownExporter is returned when the exporter passed in via the config
+	// is not a known TraceExporter type.
+	ErrUnknownExporter = errors.New("unknown tracing exporter")
+
+	defaultTimeout = time.Second * 10
+)
+
+// ConfigError is returned when there is a problem with the provided TracingConfig
+// during configuration.
+type ConfigError struct {
+	Message string
+	Err     error
+}
+
+func (c *ConfigError) Error() string {
+	if c.Err != nil {
+		return fmt.Errorf("%s, error: %w", c.Message, c.Err).Error()
+	}
+
+	return c.Message
+}
+
+// InitTracer will configure the exporter setup in the provided config and create
+// an otel TracerProvider. The new TracerProvider will be set as the global trace
+// provider.
+func InitTracer(tc Config, appName string, logger *zap.SugaredLogger) error {
+	if !tc.Enabled {
+		return nil
+	}
+
+	exp, err := configureExporter(tc)
+	if err != nil {
+		return err
+	}
+
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		// Record information about this application in an Resource.
+		sdktrace.WithResource(resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String(appName),
+			attribute.String("environment", tc.Environment),
+		)),
+	}
+
+	// exporter could be nil if we are in "passthrough" mode, but we still want
+	// to set everything up to traces go work going through the application
+	if exp != nil {
+		opts = append(opts, sdktrace.WithBatcher(exp))
+	}
+
+	tp := sdktrace.NewTracerProvider(opts...)
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+
+	return nil
+}
+
+func configureExporter(tc Config) (sdktrace.SpanExporter, error) {
```

Comment:
this is a huge function. Would it make sense to split them into `configureStdout()`, `configureJaeger()`...?